File: 10405/main.c
#include <stdio.h>
#include <memory.h>

#define MAX(a,b) ((a) > (b) ? (a) : (b))

int read_line(char* buffer)
{
   int len = 0;
   while ((*buffer = getchar()) != '\n') {
      if (*buffer == EOF) return -1;
      ++buffer;
      ++len;
   }
   *buffer = '\0';
   return len;
}

int main(int argc, char* argv[])
{
   int i, j;
   int table[1001][1001];
   char sequences[2][1001];

   memset(table, 0, sizeof(table));

   while (read_line(sequences[0]) >= 0) {
      read_line(sequences[1]);

      i = 0;
      while (sequences[0][i] != '\0') {
         j = 0;
         while (sequences[1][j] != '\0') {
            if (sequences[0][i] == sequences[1][j]) {
               table[i+1][j+1] = table[i][j] + 1;
            } else {
               table[i+1][j+1] = MAX(table[i+1][j], table[i][j+1]);
            }
            ++j;
         }
         ++i;
      }
      printf("%d\n", table[i][j]);
   }

   return 0;
}

File: 10405/Main.java
import java.io.*;

public class Main {
   public static void main(String args[]) throws IOException {
      int length;
      int[][] table = new int[1001][1001];
      String[] sequences = new String[2];
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((sequences[0] = in.readLine()) != null) {
         sequences[1] = in.readLine();
         length = 0;
         for (int i = 0; i < sequences[0].length(); i++) {
            for (int j = 0; j < sequences[1].length(); j++) {
               if (sequences[0].charAt(i) == sequences[1].charAt(j)) {
                  table[i+1][j+1] = table[i][j] + 1;
               } else {
                  table[i+1][j+1] = Math.max(table[i+1][j], table[i][j+1]);
               }
               length = table[i+1][j+1];
            }
         }
         System.out.println(length);
      }
   }
}
